### PR TITLE
[qos] use flow label to identify Very High priority IP6 packets

### PR DESCRIPTION
**Pull request by zhanglongxia**
When there are different company's devices in the same Thread Network. Some company may always set their traffic priority to High, which causes some critical traffics from other company are blocked. This feature allows the Thread device to use the flow label to identify Very High priority IP6 packets. If the devices of the same company use the same flow label, the IP6 packets with the same flow label will be set to Very High priority.

### CODE REVIEW

File: src/core/net/ip6_flow_labels.hpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+     *
+     */
+    otError GetNextFlowLabel(otIp6FlowLabelIterator &aIterator, uint32_t &aFlowLabel);
+
+private:
+    typedef struct FlowLabelEntry
+    {
+        uint32_t mFlowLabel : 20; ///< The IP6 flow label value.
+        bool     mValid : 1;      ///< Indicates whether the entry is valid.
+        uint8_t  mDelay : 8;      ///< The delay to remove flow label (in seconds).
```
Comment (bukepo) [inline src/core/net/ip6_flow_labels.hpp]: do we need allow at most 255 seconds delay?

File: src/core/net/ip6_headers.hpp [hunk 2]
```diff
@@ -210,6 +210,30 @@ class Header : private HeaderPoD
...
 
+    /**
+     * This method returns the IPv6 Flow Label value.
+     *
+     * @returns The IPv6 Flow Label value.
+     *
+     */
+    uint32_t GetFlowLabel(void) const
+    {
+        return static_cast<uint8_t>((HostSwap32(mVersionClassFlow.m32[0]) & kFlowLabelMask) >> kFlowLabelOffset);
```
Comment (bukepo) [inline src/core/net/ip6_headers.hpp]: uint8_t -> uint32_t?

File: include/openthread/message.h [hunk 3]
```diff
@@ -107,6 +107,7 @@ typedef struct otMessageSettings
 {
     bool              mLinkSecurityEnabled; ///< TRUE if the message should be secured at Layer 2.
     otMessagePriority mPriority;            ///< The message priority level.
+    uint32_t          mFlowLabel;           ///< The flow label of IP6 packet.
```
[the remark this answers is not in the document]
Comment (zhanglongxia) [reply include/openthread/message.h]: Changed.

File: src/core/common/message.hpp [hunk 4]
```diff
@@ -107,13 +107,14 @@ struct MessageInfo
...
-    bool    mInPriorityQ : 1;  ///< Indicates whether the message is queued in normal or priority queue.
-    bool    mTxSuccess : 1;    ///< Indicates whether the direct tx of the message was successful.
+    uint8_t  mType : 2;         ///< Identifies the type of message.
+    uint8_t  mSubType : 4;      ///< Identifies the message sub type.
+    bool     mDirectTx : 1;     ///< Used to indicate whether a direct transmission is required.
+    bool     mLinkSecurity : 1; ///< Indicates whether or not link security is enabled.
+    uint8_t  mPriority : 3;     ///< Identifies the message priority level (higher value is higher priority).
+    bool     mInPriorityQ : 1;  ///< Indicates whether the message is queued in normal or priority queue.
+    bool     mTxSuccess : 1;    ///< Indicates whether the direct tx of the message was successful.
+    uint32_t mFlowLabel : 20;   ///< Identifies the flow label of IP6 packet.
```
[the remark this answers is not in the document]
Comment (zhanglongxia) [reply src/core/common/message.hpp]: Changed.

File: src/core/net/ip6_headers.hpp (hunk 2 above)
Comment (zhanglongxia) [reply src/core/net/ip6_headers.hpp]: Changed.

File: src/core/net/ip6_flow_labels.hpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    {
+        kNumFlowLabelEntries = OPENTHREAD_CONFIG_IP6_FLOW_LABELS_SIZE, ///< The number of flow label entries.
+        kUpdatePeriod        = 1000,                                   ///< Update period in milliseconds.
+        kFlowLabelMask       = 0x000fffff,                             ///< The mask for flow label.
+    };
+
+    static void HandleTimer(Timer &aTimer);
+    void        HandleTimer(void);
+
+    FlowLabelEntry *FindFlowLabelEntry(uint32_t aFlowLabel, bool aValid);
```
[the remark this answers is not in the document]
Comment (zhanglongxia) [reply src/core/net/ip6_flow_labels.hpp]: Changed.

**Comment by jwhui** [comment]
Closing stale PR. Can reopen if we decide to pursue this again.

**Pull request closed without merging**